### ISSUE TITLE
Atlassian Crowd integration doesn't work after upgrading Atllasian Crowd from 3.7.1 to 4.0.0

### ISSUE THREAD

**Issue by karl-in-office** [RocketChat/Rocket.Chat#17196]
### Description:
Our Rocket.Chat server worked correctly with Atlassian Crowd 3.7.1. 
However, after upgrading Crowd from 3.7.1 to 4.0.0, Crowd integration doesn't work. 
Atlassian Crowd itself works correctly with other applications (Jira, Confluence and GitLab) after upgrading to 4.0.0. 

### Steps to reproduce:

1. Upgrade Atlassian Crowd to 4.x.
2. Open Rocket.Chat in browser and move to Administration => Atlassian Crowd.
3. Click 'Test connection'. The result says 'Invalid connectiondetails'. 

### Expected behavior:
Crowd integration works with Atlassian Crowd 4.0.x.

### Actual behavior:
Crowd integration doesn't work with Atlassian Crowd 4.0.x.

### Server Setup Information:

- Version of Rocket.Chat Server: 3.0.12
- Operating System: Red Hat Enterprise Linux Server release 7.6 (Maipo)
- Deployment Method: Manual (tar)
- Number of Running Instances: 1
- DB Replicaset Oplog: configured
- NodeJS Version: 12.14.0
- MongoDB Version: 3.6.3

### Client Setup Information

- Desktop App or Browser Version: any
- Operating System: any

### Relevant logs:

Rocket.Chat log:
`I20200407-15:10:36.568(9) server.js:204 CROWD ➔ error Invalid crowd connection details, check the url and application username/password and make sure this server is allowed to speak to crowd`

Atlassian Crowd log:
`2020-04-07 16:02:02,255 http-nio-XXXX-exec-9 INFO [plugin.rest.filter.BasicApplicationAuthenticationFilter] Invalid authentication for application with name 'APP_USERNAME'
`

**Comment by flover97** [comment]
Same problem on 3.0.12 and 3.1.0

**Comment by codemighty** [comment]
@karl-in-office, @flover97, @gabriellsh 
It happens because Crowd 4.0.0 returns `Content-Type` header as `application/json;charset=UTF-8`, but [atlassian-crowd](https://www.npmjs.com/package/atlassian-crowd) (aka [node-atlassian-crowd](https://github.com/dsn/node-atlassian-crowd)) in the method [ping](https://github.com/dsn/node-atlassian-crowd/blob/master/lib/index.js#L61) which call the method [_doRequest](https://github.com/dsn/node-atlassian-crowd/blob/master/lib/index.js#L72) expects header `Content-Type` [only as `application/json`](https://github.com/dsn/node-atlassian-crowd/blob/master/lib/index.js#L130). 

So, we just have to replace the condition from 
`if (response.headers['content-type'] !== "application/json")` 
to something like 
`if ( !(response.headers['content-type'].includes('application/json')) )`
or
`if ( !(/application\/json/.test(response.headers['content-type'])) )`
or as you wish.

**Comment by karl-in-office** [comment]
@codemighty 
Thank you for great tips!
I'll try suggested workaround in this week.

**Comment by karl-in-office** [comment]
The above workaround works with Rocket.Chat 3.1.1. 
Many thanks @codemighty !

FYI, that doesn't work with Rocket.Chat 3.2.2. 
'Test connection' returns 'Connection success', but Rocket.Chat do nothing with Atlassian Crowd. Rocket.Chat doesn't do crowd authentication and syncing name.

**Comment by karl-in-office** [comment]
Here is a log when crowd user tried to login. 
```
I20200522-20:56:34.096(9) CROWD ➔ info Init CROWD login CROWD_VALID_USER 
I20200522-20:56:34.098(9) CROWD ➔ info Extracting crowd_username 
I20200522-20:56:34.101(9) CROWD ➔ debug Local user found, redirecting to fallback login 
I20200522-20:56:34.103(9) CROWD ➔ debug User CROWD_VALID_USER is not a valid crowd user, falling back 
I20200522-20:56:34.106(9) CROWD ➔ info Fallback to default account system { username: 'CROWD_VALID_USER' } 
```

**Comment by codemighty** [comment]
> The above workaround works with Rocket.Chat 3.1.1.
> Many thanks @codemighty !

You're welcome!

> 
> FYI, that doesn't work with Rocket.Chat 3.2.2.

I'm using that workaround with Rocket.Chat 3.2.2 + Crowd 4.0.0 right now, and it works as it should; all directory's users that mapped to `rocket-chat-app` in Crowd are able to log in with no problem. I guess, the problem in your Crowd configuration or you didn't sync users' data between Rocket.Chat and Crowd. It's complicated to locate a cause without more information about your configuration.

**Comment by karl-in-office** [comment]
Really?
I'll take a look again in next week.

**Comment by karl-in-office** [comment]
> debug User CROWD_VALID_USER is not a valid crowd user, falling back

I found the root cause of the above. It is an another issue.
As far as I can remember, users who were pre-registered on Rocket.Chat could login via Crowd also.
However, it seems that such users cannot login via Crowd in recent versions.

The workaround is:
1. Ask new users to login Rocket.Chat with Crowd username/password for creating Crowd integrated accounts on Rocket.Chat. 
2. Once confirmed that all users are registered, change channel configurations (adding them, changing roles and etc.)

**Comment by atshim** [comment]
Hi, I'm a staff member under @karl-in-office .
I've tried your code with Rocket.chat 3.3.0 and it works successfully.
Thank you @codemighty :)

**Comment by karl-in-office** [comment]
Aren't there any plan to fix this issue?
It's very frustrating that we need to patch this for every Rocket.Chat update... :-(

**Comment by bvbmedia** [comment]
I can confirm the connection works properly with Crowd 2, but not with Crowd 4.

The following versions we tried:
Rocket Chat: 3.6.2
Crowd: 4.1.1

When pressing "Test connection" in admin panel the following Toastr appears:
[Invalid connection details]

When checking Rocket Chat log:
sudo service snap.rocketchat-server.rocketchat-server status
````
server.js:204 CROWD ➔ error Invalid crowd connection details, check the url and application username/password and make sure this server is allowed to speak to crowd
````

Since we use Snaps to maintain updates it is hard to manually patch the source code.

Can someone assign this issue to a Rocket Chat developer so a patch can be applied to resolve this issue?

**Comment by karl-in-office** [comment]
FYI, @codemighty 's patch is still effective for Crowd 4.1.1.

**Comment by bvbmedia** [comment]
Thanks @karl-in-office, since we use Snap the filesystem is mounted in read-only mode. So not easy to apply the patch. I read some articles but did not help me further.

Do you have any tips about how to apply a patch inside a Snap?

**Comment by bvbmedia** [comment]
Tonight I have spent some time on R&D and succeeded with applying the patch inside the Snap image.

Thanks @codemighty & @karl-in-office ! The patch is working good!

If other developers need to know how to apply a patch on their RocketChat Snap then they can read it on my blog:

https://www.basvanbeek.nl/linux/how-to-adjust-files-and-folders-inside-the-snap-file-system/

**Comment by piotrkochan** [comment]
Is it fixed in any currently available releases?

**Comment by bvbmedia** [comment]
I don't think so, because the NPM module "node-atlassian-crowd" is not yet patched:

![image](https://user-images.githubusercontent.com/3802035/94239740-88edc700-ff12-11ea-8a06-ee7328c4cf44.png)

https://github.com/dsn/node-atlassian-crowd/blob/master/lib/index.js

**Comment by piotrkochan** [comment]
I forked node-atlassian package with patch from @codemighty and it is available there:
https://www.npmjs.com/package/atlassian-crowd-patched

**Comment by karl-in-office** [comment]
> I forked node-atlassian package with patch from @codemighty and it is available there:
> https://www.npmjs.com/package/atlassian-crowd-patched

@piotrkochan 
I think that it would be better if npm information (Homepage and Repository) point to https://github.com/piotrkochan/node-atlassian-crowd

If possible, could you do PR on this repo to change package.json?
https://github.com/RocketChat/Rocket.Chat/blob/af4a21a53ddbf88045db2009384851a074183c0a/package.json#L153
Because it's difficult for me to make changes and do PR due to our company regulation :-(

**Comment by karl-in-office** [comment]
Many thanks, @piotrkochan !

**Comment by piotrkochan** [comment]
@geekgonecrazy @rodrigok please contact me, I'll give You write permission to this repository

**Comment by wysockip** [comment]
In which version we have this fixed? or it's still not fixed? (if not any estimates when it could be fixed?)

**Comment by karl-in-office** [comment]
As you can see #19025, it's still not fixed.
Hence we have to patch up manually so far. 
Please see also: https://github.com/RocketChat/Rocket.Chat/issues/17196#issuecomment-629498270

**Comment by wysockip** [comment]
Thanks for info. I've tried it - (#17196 (comment)) - but it's not working for me . Btw i'm running RC on Docker.

**Comment by codemighty** [comment]
@wysockip perhaps, your changes don't save in your Docker container or returns to default automatically. Keep in mind, you have to restart rocketchat.service after applying the workaround.

**Comment by wysockip** [comment]
@codemighty i don't think so, i've build my own rchat container (something like that):
`FROM rocketchat/rocket.chat:3.7.1`
`WORKDIR /app/bundle/programs/server/npm/node_modules/atlassian-crowd/lib`
`COPY index.js .`
`WORKDIR /app/bundle`

I end up with my-rocketchat:3.7.1 image (with fixed file mentioned here) which i then i used that to run the service, and it was not working still.

**Comment by piotrkochan** [comment]
I know that ["Open Source maintainers owe you nothing"](https://mikemcquaid.com/2018/03/19/open-source-maintainers-owe-you-nothing/) but this is just sad. This issue is present more than half year, there is known solution to this problem, there is many people who uses crowd as Rocket was heavily advertised as a replacement for HipChat. Now people want to use Rocket.Chat and it don't work for them.

**Comment by codemighty** [comment]
is it possible the condition you've replaced in `.../atlassian-crowd/lib/index.js` isn't correct, @wysockip ? or maybe you just missed `{`?

**Comment by wysockip** [comment]
@codemighty nope, i've check both your suggestions and not working
`[root@myrctest crowd_fix]# ls -lha`
`total 12K`
`drwxr-xr-x.  2 root  root    40 Dec 17 12:33 .`
`drwxr-xr-x. 15 root  root   189 Dec  9 10:25 ..`
`-rw-r--r--.  1 root  root   364 Dec 17 12:22 Dockerfile`
`-rw-r--r--.  1 65533 65533 4.8K Dec 17 12:33 index.js`
`[root@myrctest crowd_fix]# cat ./index.js | grep -i "response.headers"`
`      if ( !(/application\/json/.test(response.headers['content-type'])) ) {`

Note that i even placed the proper ownership of the file to match the one inside the docker container.

**Comment by wysockip** [comment]
by accident i found this: https://bitbucket.org/celix/celix-atlassian-crowd/commits/78d6dce91b38fd4ff980798966ad4d2c77582c31
and i think i'll give it a go

**Comment by wysockip** [comment]
I confirm that @codemighty that your fix is working fine!

I found where the problem was on my side. Believe it or not but the issue was in crowd address (missing "s").

**Comment by ITSonixcdGmbH** [comment]
is there a working solution, which does not include us creating users manually or rolling back to crowd 3.x
Rocketchat has a lot of crowd related problems. Which is not ok for a chat application which targets business.

**Comment by aduewel** [comment]
Our solution is to rebuild/modify the container before starting it:
```dockerfile
FROM rocketchat/rocket.chat:latest
RUN sed -i 's/\(response\.headers\[.content-type.\] !== "application\/json"\)/( !(response.headers["content-type"].includes("application\/json")) )/' /app/bundle/programs/server/npm/node_modules/atlassian-crowd/lib/index.js
```

**Comment by piotrkochan** [comment]
@sampaiodiego could you look at this issue?

**Comment by pierre-lehnen-rc** [comment]
This will be fixed in this month's release!

**Comment by emoxam** [comment]
It's not fixed yet ?

`2024-02-22T10:18:57+03:00 rocketchat-server.rocketchat-server[1192]: {"level":50,"time":"2024-02-22T07:18:57.697Z","pid":1192,"hostname":"rocket","name":"CROWD","msg":"Invalid crowd connection details, check the url and application username/password and make sure this server is allowed to speak to crowd","err":{"type":"Error","message":"Invalid Response from Atlassian Crowd","stack":"Error: Invalid Response from Atlassian Crowd\n    at IncomingMessage.<anonymous> (/snap/rocketchat-server/1603/programs/server/npm/node_modules/atlassian-crowd-patched/lib/index.js:131:17)\n    at IncomingMessage.emit (events.js:412:35)\n    at IncomingMessage.emit (domain.js:475:12)\n    at endReadableNT (internal/streams/readable.js:1333:12)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n => awaited here:\n    at Function.Promise.await (/snap/rocketchat-server/1603/programs/server/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:56:12)\n    at app/crowd/server/methods.ts:38:4\n    at /snap/rocketchat-server/1603/programs/server/npm/node_modules/meteor/promise/node_modules/meteor-promise/fiber_pool.js:43:40"}}`
